core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -355,12 +355,17 @@ class ZooKeeperClient(connectString: String,
   // Visibility for testing
   private[zookeeper] def scheduleSessionExpiryHandler(): Unit = {
     expiryScheduler.schedule("zk-session-expired", () => {
+      info("Session expired.")
+
+      // Initialization callbacks are invoked outside of the lock to avoid deadlock potential
+      // since callbacks may acquire their own locks and add new requests
```

Comment:
"since callbacks may acquire their own locks and add new requests" => since the completion of beforeInitializingSession() in the state change handler may require Zookeeper requests ?